platform/openiscsi/concrete_open_iscsi_admin.go

```diff
@@ -0,0 +1,162 @@
+	iscsi.logger.Debug(iscsi.logtag, "Setup Open-iscsi, initializing /etc/iscsi/initiatorname.iscsi,iscsid.conf")
+	buffer := bytes.NewBuffer([]byte{})
+	t := template.Must(template.New("Open-iscsi-initiator").Parse(initiatorNameIscsiTemplate))
+
+	type initiatorNameArgs struct {
+		Iqn string
+	}
+
+	err = t.Execute(buffer, initiatorNameArgs{iqn})
+	if err != nil {
+		err = bosherr.WrapError(err, "Generating initiatorname.iscsi of Open-iscsi")
+		return
+	}
+
+	err = iscsi.fs.WriteFile(path.Join("/etc/iscsi", "initiatorname.iscsi"), buffer.Bytes())
+	if err != nil {
+		err = bosherr.WrapError(err, "Writing to /etc/iscsi/initiatorname.iscsi")
+		return
+	}
+
+	buffer.Reset()
+
+	type iscsidConfArgs struct {
+		Username string
+		Password string
+	}
+
+	t = template.Must(template.New("Open-iscsi-conf").Parse(iscsidConfTemplate))
+	err = t.Execute(buffer, iscsidConfArgs{username, password})
+	if err != nil {
+		err = bosherr.WrapError(err, "Generating iscsid.conf of Open-iscsi")
+		return
+	}
+
+	err = iscsi.fs.WriteFile(path.Join("/etc/iscsi", "iscsid.conf"), buffer.Bytes())
+	if err != nil {
+		err = bosherr.WrapError(err, "Writing to /etc/iscsi/iscsid.conf")
+		return
+	}
+
+	buffer.Reset()
+
+	err = iscsi.Restart()
+	if err != nil {
+		err = bosherr.WrapError(err, "Restarting iscsi after modifying the /etc/iscsi/iscsid.conf file")
+		return
+	}
+
+	_, _, _, err = iscsi.runner.RunCommand("/etc/init.d/multipath-tools", "restart")
+	if err != nil {
+		err = bosherr.WrapError(err, "Restarting multipath after restarting open-iscsi")
+		return
+	}
+	return
+}
+
+// Open-iscsi initiator file - /etc/iscsi/initiatorname.iscsi
+const initiatorNameIscsiTemplate = `InitiatorName={{ .Iqn }}`
+
+// Open-iscsi deamon configuration file - /etc/iscsi/iscsid.conf
+const iscsidConfTemplate = `# Generated by bosh-agent
+node.startup = automatic
+node.session.auth.authmethod = CHAP
+node.session.auth.username = {{.Username}}
+node.session.auth.password = {{.Password}}
+discovery.sendtargets.auth.authmethod = CHAP
+discovery.sendtargets.auth.username = {{.Username}}
+discovery.sendtargets.auth.password = {{.Password}}
+node.session.timeo.replacement_timeout = 120
+node.conn[0].timeo.login_timeout = 15
+node.conn[0].timeo.logout_timeout = 15
+node.conn[0].timeo.noop_out_interval = 10
+node.conn[0].timeo.noop_out_timeout = 15
+node.session.iscsi.InitialR2T = No
+node.session.iscsi.ImmediateData = Yes
+node.session.iscsi.FirstBurstLength = 262144
+node.session.iscsi.MaxBurstLength = 16776192
+node.conn[0].iscsi.MaxRecvDataSegmentLength = 65536
+`
+
+func (iscsi ConcreteOpenIscsiAdmin) Start() (err error) {
+	iscsi.logger.Debug(iscsi.logtag, "Start Open-iscsi deamon")
+	_, _, _, err = iscsi.runner.RunCommand("/etc/init.d/open-iscsi", "start")
+	return
+}
+
+func (iscsi ConcreteOpenIscsiAdmin) Stop() (err error) {
+	iscsi.logger.Debug(iscsi.logtag, "Stop Open-iscsi deamon")
+	_, _, _, err = iscsi.runner.RunCommand("/etc/init.d/open-iscsi", "stop")
```

Comment:
In all these methods, what if there is an error? You are ignoring that and all other return values. Seems like this code is incomplete.

Also, I'd expect unit tests for each.

Reply:
Yes, these methods only return error directly. Because the cmd runner in bosh-utils could write log file entries about the running command stdout/stderr. I have referred to other code in bosh-agent. If these commands' outputs do not participate in the follow-up attendance, ignore these outputs.